apps/els_lsp/src/els_referl_unsecure_calls_diagnostics.erl

```diff
@@ -0,0 +1,137 @@
+is_default() ->
+  true.
+
+-spec run(uri()) -> [els_diagnostics:diagnostic()].
+run(Uri)->
+  run(Uri, 0).
+
+
+-spec run(uri(), number()) -> [els_diagnostics:diagnostic()]. 
+run(Uri, RecursionDepth) when RecursionDepth < ?MAX_RECURSION_DEPTH ->
+  case filename:extension(Uri) of
+    <<".erl">> -> 
+      case referl_node() of
+        disabled ->
+          [];
+        Node -> 
+          case rpc:call(Node, ri, add, [binary_to_list(els_uri:path(Uri))], ?TIME_OUT) of
+            {badrpc, _} ->
+              els_server:send_notification(<<"window/showMessage">>, #{ type => ?MESSAGE_TYPE_ERROR, message => <<"Refactor Erl node is down {badrpc}!">> }),
+              [];
+            error ->
+              timer:sleep(1000),
+              run(Uri, RecursionDepth + 1);
+            _ ->
+              ModuleName = filename:rootname(filename:basename(binary_to_list(els_uri:path(Uri)))),
+              ReferlResult = rpc:call(Node, refusr_sq, run, [[{positions, linecol}, {output, msg}], [], "mods[name=" ++ ModuleName ++ "].funs.unsecure_calls"]),
+              Pois = convertToPoi(ReferlResult),
+              [make_diagnostic(Poi) || Poi <- Pois]
+          end
+      end;
+    _ ->
+      []
+  end;
+
+run(_, RecursionDepth) when RecursionDepth >= ?MAX_RECURSION_DEPTH ->
+  els_server:send_notification(<<"window/showMessage">>, #{ type => ?MESSAGE_TYPE_ERROR, message => <<"Cannot add module to RefactorErl!">> }),
+  [].
+
+-spec source() -> binary().
+source() ->
+  <<"Unsecure Calls (Referl)">>.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+-spec convertToPoi(any()) -> poi(). 
+convertToPoi(ReferlResult) ->
+case ReferlResult of
+    [{_, _, _, DataList}] -> % [{Option, Tuple, Atom, DataList}] = Out
+      convertToPoi(DataList);
+    [{{_, {FromLine, FromCol}, {ToLine, ToCol}}, Name} | Tail] -> %  [ {{Path, StartPos, EndPos}, MacroName} | Tail] = DataList
+      Range = #{ from => {FromLine, FromCol}, to => {ToLine, ToCol} },
+      Id = referl_atom_unsec, %"{module(), 'atom()', 'arity()''}",
+      Poi = els_poi:new(Range, application, Id, Name), % Additional Data param can be added
+      [ Poi | convertToPoi(Tail) ];
+    _ ->
+        [] %TODO Robi notify
+end.
+
+-spec make_diagnostic(poi()) -> els_diagnostics:diagnostic().
+make_diagnostic(#{ data := PoiData, range := PoiRange}) -> %#{id := POIId, range := POIRange} .   
+    Range = els_protocol:range(PoiRange),
+    IssueName = list_to_binary(PoiData), % TODO Robi: the ID and the arity should be encoded to the ID
+    Message = <<"Security Issue: ", IssueName/binary>>,
+    Severity = ?DIAGNOSTIC_WARNING,
+    Source = source(),
+    els_diagnostics:make_diagnostic(Range, Message, Severity, Source).
+
+  
+-spec referl_node() -> atom(). % If called with no args, then try to get it from config
+referl_node() ->
```

Comment:
looks like some common referl functionality could be moved to a library module like `els_referl_util`, eg `referl_node/0/1`, error handling of rpc call to the referl node, processing common referl responses like `[{Option, Tuple, Atom, DataList}]` with a fun etc

Reply:
@gomoripeti I just started an els_referl module for the "common parts" should be finished by the end of the week. Also cleaned out refactor erl node detectiona and error management.

Reply:
awesome

Reply:
I agree. It would be nice to decouple the diagnostic backend itself from a module which deals with interfacing with RefactorErl. These could be a single `els_refactorerl_diagnostics.erl` (which covers all, configurable, diagnostic types) and `els_refactorerl.erl`